File: app/src/main/java/com/wireguard/android/model/Tunnel.java
/*
 * Copyright © 2018 Samuel Holland <samuel@sholland.org>
 * Copyright © 2018 Jason A. Donenfeld <Jason@zx2c4.com>. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

package com.wireguard.android.model;

import androidx.annotation.Nullable;

import com.wireguard.android.config.Config;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import net.ivpn.client.IVPNApplication;

public class Tunnel {
    private static final Logger LOGGER = LoggerFactory.getLogger(Tunnel.class);

    private Config config;
    private String name;
    private State state;

    public Tunnel(final String name, final Config config, final State state) {
        this.name = name;
        this.config = config;
        this.state = state;
    }

    @Nullable
    public Config getConfig() {
        return config;
    }

    public String getName() {
        return name;
    }

    public State getState() {
        return state;
    }

    public void setName(String name) {
        this.name = name;
    }

    public void setState(final State state) {
        try {
            this.state = state;
            IVPNApplication.getApplication().appComponent.provideGoBackend().setState(this, state);
        } catch (Exception e) {
            LOGGER.error(e.getLocalizedMessage());
        }
    }

    public enum State {
        DOWN,
        TOGGLE,
        UP;

        public static State of(final boolean running) {
            return running ? UP : DOWN;
        }
    }
}

File: app/src/main/java/com/wireguard/android/model/Tunnel.kt
/*
 * Copyright © 2018 Samuel Holland <samuel@sholland.org>
 * Copyright © 2018 Jason A. Donenfeld <Jason@zx2c4.com>. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
package com.wireguard.android.model

import com.wireguard.android.config.Config
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import net.ivpn.client.IVPNApplication
import org.slf4j.LoggerFactory

class Tunnel(var name: String, var config: Config?, private var state: State) {

    var listener: OnStateChangedListener? = null

    fun getState(): State {
        return state
    }

    suspend fun setState(state: State) {
        try {
            this.state = state
            LOGGER.info("Set state = $state")
            withContext(Dispatchers.IO) {
                IVPNApplication.getApplication().appComponent.provideGoBackend().setState(this@Tunnel, state, config)
            }
        } catch (e: Exception) {
            LOGGER.error(e.localizedMessage)
        }
    }

    fun onStateChange(newState: State) {
        GlobalScope.launch {
            withContext(Dispatchers.Main) {
                listener?.onStateChanged(newState)
            }
        }
//        listener?.onStateChanged(newState)
    }

    enum class State {
        DOWN, TOGGLE, UP;

        companion object {
            fun of(running: Boolean): State {
                return if (running) UP else DOWN
            }
        }
    }

    companion object {
        private val LOGGER = LoggerFactory.getLogger(Tunnel::class.java)
    }

    interface OnStateChangedListener {
        fun onStateChanged(newState: State)
    }
}